src/main.rs

```diff
@@ -68,96 +79,275 @@ async fn main() {
-                (None, true, true) => &EntryOutput::Both,
-                (None, false, false) => &EntryOutput::Id, // default mode
+                (Some(e), false, false) => *e,
+                (None, true, false) => EntryOutput::Id,
+                (None, false, true) => EntryOutput::Path,
+                (None, true, true) => EntryOutput::Both,
+                (None, false, false) => EntryOutput::Link, // default value
                 _ => panic!(
                     "incompatible entry output options, please choose only one"
                 ),
             };
 
-            let index = provide_index(&root).expect("could not provide index");
-
-            let resource_index = index.read().expect("could not read index");
-
-            let mut resources = resource_index
+            let mut storage_entries: Vec<StorageEntry> = provide_index(&root)
+                .expect("could not provide index")
+                .read()
+                .expect(
+                    "
+            could not read index",
+                )
                 .path2id
                 .iter()
                 .map(|(path, resource)| {
-                    let tags_list = read_storage_value(
-                        &root,
-                        "tags",
-                        &resource.id.to_string(),
-                        &None,
-                    )
-                    .unwrap_or("NO_TAGS".to_string());
-
-                    let scores_list = read_storage_value(
-                        &root,
-                        "scores",
-                        &resource.id.to_string(),
-                        &None,
-                    )
-                    .unwrap_or("NO_SCORE".to_string());
-
-                    let datetime = DateTime::<Utc>::from(resource.modified);
-
-                    (path, resource, tags_list, scores_list, datetime)
+                    let tags = if *tags {
+                        Some(
+                            read_storage_value(
+                                &root,
+                                "tags",
+                                &resource.id.to_string(),
+                                &None,
+                            )
+                            .map_or(vec![], |s| {
+                                s.split(',')
+                                    .map(|s| s.trim().to_string())
+                                    .collect::<Vec<_>>()
+                            }),
+                        )
+                    } else {
+                        None
+                    };
+
+                    let scores = if *scores {
+                        Some(
+                            read_storage_value(
+                                &root,
+                                "scores",
+                                &resource.id.to_string(),
+                                &None,
+                            )
+                            .map_or(0, |s| s.parse::<u32>().unwrap_or(0)),
```

Comment:
We should add something like `typealias Score` into `ark-rust` repo.

Not sure how could we name the crate, maybe `metadata-score`?

Reply:
We could also modify `ResourceIndex` to hold this information
For example, 
```rust
#[derive(Debug, Clone, PartialEq, Eq, Hash, Ord, PartialOrd)]
pub enum TagType {
    Tag(String),
    Score(String),
}

#[derive(Eq, Ord, PartialEq, PartialOrd, Hash, Clone, Debug)]
pub struct IndexEntry {
    pub modified: SystemTime,
    pub id: ResourceId,
    pub tags: Option<Vec<TagType>>,
}

#[derive(PartialEq, Clone, Debug)]
pub struct ResourceIndex {
    pub id2path: HashMap<ResourceId, CanonicalPathBuf>,
    pub path2id: HashMap<CanonicalPathBuf, IndexEntry>,

    pub collisions: HashMap<ResourceId, usize>,
    root: PathBuf,
}
```

`tags` here are optional so it''s backward compatible.

Reply:
@tareknaser no, I would not mix metadata into the index. The storages are made separate intentionally, because they are specific to use-case. Otherwise the index structure would bloat with too many stuff which not everyone needs..

Reply:
I can imagine building some similar structure to your proposed `IndexEntry` but in the end of the pipeline and only if we could figure out dynamically fields in it.

The generic "pipeline" for users of `ark-rust` would follow this flow:
```
[ build index ] -> [ generate metadata ] -> [ rich entries ]
           |----------> [ scan storages ] ----------------^
```

At this moment, in Kotlin version of our little framework we have `[ generate metadata ]` stage, which extracts some metadata and also generates previews (which is like graphical metadata), both are stored in cache storages. The `[ scan storages ]` stage is implicit yet, the user queries storages by ids, but we could implement it and construct "rich entries" in the end with all relevant metadata.

Reply:
> The storages are made separate intentionally, because they are specific to use-case. Otherwise the index structure would bloat with too many stuff which not everyone needs..


If this use-case isn't typical for the user, then I agree it shouldn't be the default and stored as part of the index.

> The [ scan storages ] stage is implicit yet, the user queries storages by ids, but we could implement it and construct "rich entries" in the end with all relevant metadata.

Rich entries seem like a good idea. We could also define another type for them so to avoid confusion.